framework/src/meshgenerators/MeshGenerator.C

```diff
@@ -135,15 +238,52 @@ MeshGenerator::generateInternal()
   return mesh;
 }
 
-std::unique_ptr<MeshBase> &
-MeshGenerator::addMeshSubgenerator(const std::string & generator_name,
+void
+MeshGenerator::addMeshSubgenerator(const std::string & type,
                                    const std::string & name,
-                                   InputParameters & params)
+                                   InputParameters params)
 {
+  if (!_app.constructingMeshGenerators())
+    mooseError("Can only call addMeshSubgenerator() during MeshGenerator construction");
+
   // In case the user forgot it
   params.set<MooseApp *>("_moose_app") = &_app;
 
-  _app.addMeshGenerator(generator_name, name, params);
+  _app.addMeshGenerator(type, name, params);
+  _sub_mesh_generators.insert(&std::as_const(_app).getMeshGenerator(name));
+}
+
+RestartableDataValue &
+MeshGenerator::setMeshPropertyHelper(const std::string & data_name)
+{
+  return _app.getRestartableMetaData(meshPropertyName(data_name), MooseApp::MESH_META_DATA, 0);
+}
+
+void
+MeshGenerator::addParentMeshGenerator(const MeshGenerator & mg, const AddParentChildKey)
+{
+  _parent_mesh_generators.insert(&mg);
+}
 
-  return this->getMeshByName(name);
+void
+MeshGenerator::addChildMeshGenerator(const MeshGenerator & mg, const AddParentChildKey)
+{
+  _child_mesh_generators.insert(&mg);
+}
+
+bool
+MeshGenerator::isParentMeshGenerator(const MeshGeneratorName & name) const
+{
+  return std::find_if(getParentMeshGenerators().begin(),
+                      getParentMeshGenerators().end(),
+                      [&name](const auto & mg)
+                      { return mg->name() == name; }) != getParentMeshGenerators().end();
+}
+
+void
+MeshGenerator::declareNullMeshName(const MeshGeneratorName & name)
+{
+  mooseAssert(_app.constructingMeshGenerators(), "Should only be called at construction");
+  mooseAssert(!_null_mesh_names.count(name), "Already declared");
```

Comment:
would remove that one

I can see MGs where dummies would be needed in two separate inputs and they could be passed by a user with the same name

Reply:
not sure this makes any sense to me

Reply:
We could have two inputs called dummy in two separate parameters of an object

Reply:
This is actually supported and has nothing to do with the declaration. I'll add a test for it